arnoldPlugins/gaffer.mtd

```diff
@@ -199,79 +203,102 @@
 
 	gaffer.layout.section.Shape.collapsed BOOL false
 	gaffer.layout.section.Sampling.collapsed BOOL false
+	gaffer.layout.activator.modeIsPortal STRING "parameters['portal'].getValue() == False"
 
 	[attr portal]
 		options STRING "Light:0|Portal:1"
 		label STRING "Mode"
 		gaffer.layout.index INT 0
 	[attr color]
 		gaffer.layout.index INT 1
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr intensity]
 		gaffer.layout.index INT 2
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr exposure]
 		gaffer.layout.index INT 3
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr normalize]
 		gaffer.layout.index INT 4
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr aov]
 		gaffer.layout.index INT 5
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 
 	[attr roundness]
 		page STRING "Shape"
 		gaffer.layout.index INT 6
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr soft_edge]
 		page STRING "Shape"
 		gaffer.layout.index INT 7
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr spread]
 		page STRING "Shape"
 		gaffer.layout.index INT 8
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr resolution]
 		page STRING "Shape"
 		gaffer.layout.index INT 9
 
 	[attr samples]
 		page STRING "Sampling"
 		gaffer.layout.index INT 10
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr volume_samples]
 		page STRING "Sampling"
 		gaffer.layout.index INT 11
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 
 	[attr cast_shadows]
 		page STRING "Shadows"
 		gaffer.layout.index INT 12
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr cast_volumetric_shadows]
 		page STRING "Shadows"
 		gaffer.layout.index INT 13
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr shadow_density]
 		page STRING "Shadows"
 		gaffer.layout.index INT 14
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr shadow_color]
 		page STRING "Shadows"
 		gaffer.layout.index INT 15
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 
 	[attr camera]
 		page STRING "Contribution"
 		gaffer.layout.index INT 16
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr transmission]
 		page STRING "Contribution"
 		gaffer.layout.index INT 17
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr diffuse]
 		page STRING "Contribution"
 		gaffer.layout.index INT 18
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr specular]
 		page STRING "Contribution"
 		gaffer.layout.index INT 19
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr sss]
 		page STRING "Contribution"
 		gaffer.layout.index INT 20
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
 	[attr indirect]
 		page STRING "Contribution"
 		gaffer.layout.index INT 21
+		gaffer.layout.visibilityActivator STRING "modeIsPortal"
```

Comment:
It's a pity that we're leaving behind these empty "Contribution" etc sections. If I recall correctly, this is also the reason we didn't add filetype-based activators to the ImageWriter format sections. Worth a quick look in PlugLayout to see if we can automatically hide a section if all its children are hidden?

Reply:
1ef4f995038121cf06e701dbf8549ecf0280ee15 hides sections with no visible children. I think without any side effects. I also took a stab at making `ImageWriter` a little cleaner in dad9f5b622b273e68be9a5832a5a993aecbc563a and it feels like a good improvement.

Reply:
Nice! I do worry a little that there might be scenarios where `ImageWriter.fileName` is driven by an expression that can't determine a filename in the main UI context (it depends on a wedge variable or something like that). In that case, I think we'd probably want to show all the sections just in case. I think the logic might be a bit fiddly, maybe something like this?

```
def __extension( plug ) :

    fileNamePlug = plug.parent["fileName"]
    fileName = fileNamePlug.getValue()
    ext = os.path.splitext( fileName )[1]
    if ext :
        return ext.lower()[1:]
    return "" if fileNamePlug.isSetToDefault() else "unknown"

"layout:activator:dpx", lambda p : __extension( p ) in ( "dpx", "unknown" )
```

Reply:
Oh, and this isn't completely what was requested in https://github.com/GafferHQ/gaffer/issues/1477, but I think it's close enough that we can close it with `Fixes #1477` in the commit message.

Reply:
Good point, fixed in the squashed commit at b3aa336d3609c09f4d612129cce988eca5e5eb40.

I also closed #1477 with a note pointing to this PR.